### PR TITLE
[AMDGPU][MC][True16] VOP3dot instruction update for true16/fake16

### PR DESCRIPTION
**Pull request by broxigarchen**
Update VOP3dot instructions with true16 and fake16 formats.

This patch includes instructions:
v_dot2_f16_f16
v_dot2_bf16_bf16

### CODE REVIEW

**Review by arsenm — APPROVED** [review]

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s [hunk 1]
```diff
@@ -2119,104 +2119,116 @@ v_div_scale_f64 v[254:255], vcc, 0xaf123456, -vcc, -1 clamp div:2
...
 
-v_dot2_bf16_bf16 v2, 0x20004000, v0, v2
-// GFX11: v_dot2_bf16_bf16 v2, 0x20004000, v0, v2 ; encoding: [0x02,0x00,0x67,0xd6,0xff,0x00,0x0a,0x04,0x00,0x40,0x00,0x20]
+v_dot2_bf16_bf16 v2.l, 0x20004000, v0, v2.l
+// GFX11: [0x02,0x00,0x67,0xd6,0xff,0x00,0x0a,0x04,0x00,0x40,0x00,0x20]
 
-v_dot2_f16_f16 v5, v1, v2, s3
-// GFX11: encoding: [0x05,0x00,0x66,0xd6,0x01,0x05,0x0e,0x00]
+v_dot2_bf16_bf16 v5.l, vcc_lo, ttmp15, v255.h
+// GFX11: [0x05,0x20,0x67,0xd6,0x6a,0xf6,0xfc,0x07]
```
Comment (kosarev) [inline llvm/test/MC/AMDGPU/gfx11_asm_vop3.s:2171]: If we don't want the 'encoding:' bits here, please remove them in the whole file with a separate patch (or maybe you could just add them back downstream). This should hopefully help the diff algorithms to show additions as such and not as reshuffled instructions as in this case. This will also likely help merging the change whenever it doesn't match downstream exactly.

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt [hunk 2]
```diff
@@ -3395,32 +3395,92 @@
...
+# W64-FAKE16: v_dot2_bf16_bf16_e64_dpp v255, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x47,0x67,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00]
+0xff,0x47,0x67,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00
+
+# GFX11: v_dot2_bf16_bf16_e64_dpp v5.l, v1, v2, v255.h op_sel:[0,0,1,0] dpp8:[7,6,5,4,3,2,1,0] ; encoding: [0x05,0x20,0x67,0xd6,0xe9,0x04,0xfe,0x07,0x01,0x77,0x39,0x05]
+0x05,0x20,0x67,0xd6,0xe9,0x04,0xfe,0x07,0x01,0x77,0x39,0x05
+
+# GFX11: v_dot2_bf16_bf16_e64_dpp v255.h, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x47,0x67,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00]
+0xff,0x47,0x67,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00
+
+# op_sel[1:0] are ignored
```
Comment (kosarev) [inline llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt:3483]: What does this comment refer to and why? op_sel seems to be 0 in this code.

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt [hunk 3]
```diff
@@ -3395,32 +3395,92 @@
...
+# W64-FAKE16: v_dot2_f16_f16_e64_dpp v255, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x47,0x66,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00]
+0xff,0x47,0x66,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00
+
+# GFX11: v_dot2_f16_f16_e64_dpp v5.l, v1, v2, v255.h op_sel:[0,0,1,0] dpp8:[7,6,5,4,3,2,1,0] ; encoding: [0x05,0x20,0x66,0xd6,0xe9,0x04,0xfe,0x07,0x01,0x77,0x39,0x05]
+0x05,0x20,0x66,0xd6,0xe9,0x04,0xfe,0x07,0x01,0x77,0x39,0x05
+
+# GFX11: v_dot2_f16_f16_e64_dpp v255.h, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x47,0x66,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00]
+0xff,0x47,0x66,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00
 
 # op_sel[1:0] are ignored
```
Comment (kosarev) [inline llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt:3440]: What does this comment refer to and why? op_sel seems to be 0 in this code.

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s [hunk 4]
```diff
@@ -2119,104 +2119,116 @@ v_div_scale_f64 v[254:255], vcc, 0xaf123456, -vcc, -1 clamp div:2
...
 // GFX11: encoding: [0x05,0x06,0x66,0xd6,0xc1,0xfe,0xf4,0xc3]
 
-v_dot2_f16_f16 v5, 0.5, -m0, 0.5 op_sel:[0,0,0,0]
+v_dot2_f16_f16 v5.l, 0.5, -m0, 0.5
 // GFX11: encoding: [0x05,0x00,0x66,0xd6,0xf0,0xfa,0xc0,0x43]
 
-v_dot2_f16_f16 v5, -src_scc, |vcc_lo|, -1 op_sel:[0,0,1,0]
-// GFX11: encoding: [0x05,0x22,0x66,0xd6,0xfd,0xd4,0x04,0x23]
+v_dot2_f16_f16 v5.l, -src_scc, |vcc_lo|, -1
+// GFX11: encoding: [0x05,0x02,0x66,0xd6,0xfd,0xd4,0x04,0x23]
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s:2216]: I believe the old op_sel testline is selecting .h on the -1 and the assembely does matched with it

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s [hunk 5]
```diff
@@ -4700,44 +4700,117 @@ v_pack_b32_f16_e64_dpp v5, -v1, |v2| op_sel:[1,0,0] row_xmask:0 row_mask:0x1 ban
 v_pack_b32_f16_e64_dpp v255, -|v255|, -|v255| op_sel:[0,1,0] row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:1 fi:1
 // GFX11: [0xff,0x13,0x11,0xd7,0xfa,0xfe,0x03,0x60,0xff,0x6f,0x0d,0x30]
 
-v_dot2_f16_f16_e64_dpp v0, v1, v2, v3 quad_perm:[0,1,2,3] row_mask:0x0 bank_mask:0x0 fi:1
-// GFX11: encoding: [0x00,0x00,0x66,0xd6,0xfa,0x04,0x0e,0x04,0x01,0xe4,0x04,0x00]
+v_dot2_f16_f16_e64_dpp v0.l, v1, v2, v3.l op_sel:[1,1,0,0] quad_perm:[0,1,2,3] row_mask:0x0 bank_mask:0x0 fi:1
+// GFX11-ERR: :[[@LINE-1]]:{{[0-9]+}}: error: invalid op_sel operand
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s:4704]: I think there are some testline missing for op_sel testlines, and we might have same problem in this https://github.com/llvm/llvm-project/pull/113603 as well. I will check and address these

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt [hunk 6]
```diff
@@ -1983,105 +1983,228 @@
 # W64: v_div_scale_f64 v[254:255], vcc, 0xaf123456, -vcc, -1 clamp div:2 ; encoding: [0xfe,0xea,0xfd,0xd6,0xff,0xd4,0x04,0x5b,0x56,0x34,0x12,0xaf]
 0xfe,0xea,0xfd,0xd6,0xff,0xd4,0x04,0x5b,0x56,0x34,0x12,0xaf
 
-# GFX11: v_dot2_bf16_bf16 v5, v1, v2, s3         ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0x0e,0x00]
+# W32-REAL16: v_dot2_bf16_bf16 v5.l, v1, v2, s3       ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0x0e,0x00]
+# W32-FAKE16: v_dot2_bf16_bf16 v5, v1, v2, s3         ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0x0e,0x00]
+# W64-REAL16: v_dot2_bf16_bf16 v5.l, v1, v2, s3       ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0x0e,0x00]
+# W64-FAKE16: v_dot2_bf16_bf16 v5, v1, v2, s3         ; encoding: [0x05,0x00,0x67,0xd6,0x01,0x05,0x0e,0x00]
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3.txt:1989]: I think we can keep it as it for now since fake16 testline will be removed later. I can merge those in the later patch

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt (hunk 3 above)
Comment (Sisyph) [reply llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt:3440]: The comment refers to the actual semantics of the instructions, in that the opsel bit for src0 and src1 is ignored by hardware. And we can see from the test those opsel bits are also stripped out by the disassembler. The input bytes set those bits, and the output disassembly does not set them. Yes this is a problem for round tripping. It looks like a problem with the script moving comments around, and they should be put back on the instructions they pertain to (it looks like those where the second byte of input is 0x78).

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sisyph) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s:2216]: Raising a .h bit on an non-VGPR operand should be valid. We should keep the existing test cases (which requires adding .l to the dst operand, to match the op_sel operand).

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s [hunk 7]
```diff
@@ -4700,44 +4700,117 @@ v_pack_b32_f16_e64_dpp v5, -v1, |v2| op_sel:[1,0,0] row_xmask:0 row_mask:0x1 ban
...
+
+v_dot2_f16_f16_e64_dpp v0.l, s1, v2, v3.l quad_perm:[0,1,2,3] row_mask:0x0 bank_mask:0x0 fi:1
+// GFX11-ERR: :[[@LINE-1]]:{{[0-9]+}}: error: invalid operand for instruction
+
+v_dot2_f16_f16_e64_dpp v0.l, v1, s2, v3.l quad_perm:[0,1,2,3] row_mask:0x0 bank_mask:0x0 fi:1
+// GFX11-ERR: :[[@LINE-1]]:{{[0-9]+}}: error: invalid operand for instruction
 
-v_dot2_f16_f16_e64_dpp v0, v1, v2, v3 op_sel:[1,1,0,0] quad_perm:[0,1,2,3] row_mask:0x0 bank_mask:0x0 fi:1
+v_dot2_bf16_bf16_e64_dpp v0.l, v1, v2, v3.l op_sel:[1,1,0,0] quad_perm:[0,1,2,3] row_mask:0x0 bank_mask:0x0 fi:1
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s]: changed the order

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt (hunk 2 above)
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt]: moved the right place

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s (hunk 5 above)
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s:4707]: added back the [0x00,0x00,0x66,0xd6,0xfa,0x04,0x0e,0x04,0x01,0xe4,0x04,0x00] line, and also added some missing op_sel testlines

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt [hunk 8]
```diff
@@ -3395,32 +3395,92 @@
...
+# W32-FAKE16: v_dot2_bf16_bf16_e64_dpp v255, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x47,0x67,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00]
+# W64-REAL16: v_dot2_bf16_bf16_e64_dpp v255.h, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x47,0x67,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00]
+# W64-FAKE16: v_dot2_bf16_bf16_e64_dpp v255, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x47,0x67,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00]
+0xff,0x47,0x67,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00
+
+# GFX11: v_dot2_bf16_bf16_e64_dpp v5.l, v1, v2, v255.h op_sel:[0,0,1,0] dpp8:[7,6,5,4,3,2,1,0] ; encoding: [0x05,0x20,0x67,0xd6,0xe9,0x04,0xfe,0x07,0x01,0x77,0x39,0x05]
+0x05,0x20,0x67,0xd6,0xe9,0x04,0xfe,0x07,0x01,0x77,0x39,0x05
+
+# GFX11: v_dot2_bf16_bf16_e64_dpp v255.h, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] dpp8:[0,0,0,0,0,0,0,0] fi:1 ; encoding: [0xff,0x47,0x67,0xd6,0xea,0xfe,0xf7,0xe3,0xff,0x00,0x00,0x00]
```
Comment (Sisyph) [inline llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt:3481]: GFX11 in a gfx12 file, something is wrong.

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sisyph) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s:2216]: It looks like we are still missing tests that existed previously. Where is this test? "v_dot2_f16_f16 v5, -src_scc, |vcc_lo|, -1 op_sel:[0,0,1,0]"

The other asm test files we didn't lose any tests I could see. Maybe it's just this one?

[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s]: Add back op_sel to this test. Also for gfx12 as well

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt (hunk 8 above)
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp8.txt]: Fixed

**Review by Sisyph — APPROVED** [review]
LGTM

**Comment by broxigarchen** [comment]
Squash the commits for the ease of bringing up downstreaming patch

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s [hunk 9]
```diff
@@ -4700,44 +4700,132 @@ v_pack_b32_f16_e64_dpp v5, -v1, |v2| op_sel:[1,0,0] row_xmask:0 row_mask:0x1 ban
...
+// GFX11: encoding: [0xff,0x47,0x67,0xd6,0xfa,0xfe,0xf7,0xe3,0xff,0x6f,0x05,0x30]
+
+v_dot2_bf16_bf16_e64_dpp v5.l, v1, v2, v255.h row_mirror
+// GFX11: encoding: [0x05,0x20,0x67,0xd6,0xfa,0x04,0xfe,0x07,0x01,0x40,0x01,0xff]
+
+v_dot2_bf16_bf16_e64_dpp v255.h, -|v255|, -|v255|, -|src_scc| row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:0 fi:1
+// GFX11: encoding: [0xff,0x47,0x67,0xd6,0xfa,0xfe,0xf7,0xe3,0xff,0x6f,0x05,0x30]
 
-v_dot2_bf16_bf16_e64_dpp v5, v1, v2, 0 quad_perm:[3,2,1,0] row_mask:0xf bank_mask:0xf
-// GFX11: encoding: [0x05,0x00,0x67,0xd6,0xfa,0x04,0x02,0x02,0x01,0x1b,0x00,0xff]
```
Comment (kosarev) [inline llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s:4743]: Why is this removed?

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp8.s [hunk 10]
```diff
@@ -3009,44 +3009,126 @@ v_pack_b32_f16_e64_dpp v5, -v1, |v2| op_sel:[1,0,0] dpp8:[7,6,5,4,3,2,1,0]
...
 
-v_dot2_f16_f16_e64_dpp v0, |v1|, -v2, -|s3| op_sel:[0,0,1,1] dpp8:[0,1,2,3,4,4,4,4]
-// GFX11: encoding: [0x00,0x65,0x66,0xd6,0xe9,0x04,0x0e,0xc0,0x01,0x88,0x46,0x92]
+v_dot2_f16_f16_e64_dpp v5.l, v1, v2, v255.h dpp8:[7,6,5,4,3,2,1,0]
+// GFX11: encoding: [0x05,0x20,0x66,0xd6,0xe9,0x04,0xfe,0x07,0x01,0x77,0x39,0x05]
 
-v_dot2_f16_f16_e64_dpp v5, v1, v2, 0.5 dpp8:[7,6,5,4,3,2,1,0]
-// GFX11: encoding: [0x05,0x00,0x66,0xd6,0xe9,0x04,0xc2,0x03,0x01,0x77,0x39,0x05]
+v_dot2_f16_f16_e64_dpp v5.l, v1, v2, s3 dpp8:[7,6,5,4,3,2,1,0]
+// GFX11: encoding: [0x05,0x00,0x66,0xd6,0xe9,0x04,0x0e,0x00,0x01,0x77,0x39,0x05]
```
Comment (kosarev) [inline llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp8.s:3034]: I don't know how to review such changes. It seems to replace the upstream version of the file with its downstream version, which look to be heavily out of sync.

How do we know we don't drop any useful cases here?

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3.s (hunk 4 above)
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3.s]: resolved test divergence. All tests and order are maintained

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp8.s (hunk 10 above)
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp8.s]: same as above

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s (hunk 9 above)
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16.s]: resolve test divergenced

**Review by arsenm — APPROVED** [review]

**Review by mariusz-sikora-at-amd — COMMENTED** [review]
This is not a full list, but some tests are duplicates

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp16.txt [hunk 11]
```diff
@@ -5445,30 +5445,102 @@
...
+
+0xff,0x47,0x66,0xd6,0xfa,0xfe,0xf7,0xe3,0xff,0x6f,0x0d,0x30
+# W32-REAL16: v_dot2_f16_f16_e64_dpp v255.h, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:1 fi:1 ; encoding: [0xff,0x47,0x66,0xd6,0xfa,0xfe,0xf7,0xe3,0xff,0x6f,0x0d,0x30]
+# W32-FAKE16: v_dot2_f16_f16_e64_dpp v255, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:1 fi:1 ; encoding: [0xff,0x47,0x66,0xd6,0xfa,0xfe,0xf7,0xe3,0xff,0x6f,0x0d,0x30]
+# W64-REAL16: v_dot2_f16_f16_e64_dpp v255.h, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:1 fi:1 ; encoding: [0xff,0x47,0x66,0xd6,0xfa,0xfe,0xf7,0xe3,0xff,0x6f,0x0d,0x30]
+# W64-FAKE16: v_dot2_f16_f16_e64_dpp v255, -|v255|, -|v255|, -|src_scc| op_sel:[0,0,0,1] row_xmask:15 row_mask:0x3 bank_mask:0x0 bound_ctrl:1 fi:1 ; encoding: [0xff,0x47,0x66,0xd6,0xfa,0xfe,0xf7,0xe3,0xff,0x6f,0x0d,0x30]
+
+0x05,0x20,0x66,0xd6,0xfa,0x04,0xfe,0x07,0x01,0xe4,0x00,0xff
```
[the remark this answers is not in the document]
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_dpp16.txt:5484]: Thanks for pointing this out. I will bring up a patch to run "--unique" on these files
